x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingResourceTracker.java

```diff
@@ -0,0 +1,160 @@
+        MlMemoryTracker mlMemoryTracker,
+        ActionListener<MlAutoscalingStats> listener
+    ) {
+        String[] mlNodes = clusterState.nodes()
+            .stream()
+            .filter(node -> node.getRoles().contains(DiscoveryNodeRole.ML_ROLE))
+            .map(DiscoveryNode::getId)
+            .toArray(String[]::new);
+
+        getMlNodeStats(
+            mlNodes,
+            client,
+            timeout,
+            ActionListener.wrap(
+                osStatsPerNode -> getMemoryAndCpu(clusterState, mlMemoryTracker, osStatsPerNode, listener),
+                listener::onFailure
+            )
+        );
+
+    }
+
+    static void getMlNodeStats(String[] mlNodes, Client client, TimeValue timeout, ActionListener<Map<String, OsStats>> listener) {
+        client.admin()
+            .cluster()
+            .prepareNodesStats(mlNodes)
+            .clear()
+            .setOs(true)
+            .setTimeout(timeout)
+            .execute(
+                ActionListener.wrap(
+                    nodesStatsResponse -> listener.onResponse(
+                        nodesStatsResponse.getNodes()
+                            .stream()
+                            .collect(Collectors.toMap(nodeStats -> nodeStats.getNode().getId(), NodeStats::getOs))
+                    ),
+                    listener::onFailure
+                )
+            );
+    }
+
+    static void getMemoryAndCpu(
+        ClusterState clusterState,
+        MlMemoryTracker mlMemoryTracker,
+        Map<String, OsStats> osStatsPerNode,
+        ActionListener<MlAutoscalingStats> listener
+    ) {
+        Set<String> nodesWithRunningJobs = new HashSet<>();
+        long memoryBytesSum = osStatsPerNode.values()
+            .stream()
+            .map(s -> s.getMem().getAdjustedTotal().getBytes())
+            .collect(Collectors.summingLong(Long::longValue));
+
+        long modelMemoryBytesSum = 0;
+        long extraSingleNodeModelMemoryInBytes = 0;
+        int extraSingleNodeProcessors = 0;
+        int minNodes = 0;
+        int extraProcessors = 0;
+
+        final MlAutoscalingContext autoscalingContext = new MlAutoscalingContext(clusterState);
+
+        for (var task : autoscalingContext.anomalyDetectionTasks) {
+            String jobId = ((OpenJobAction.JobParams) task.getParams()).getJobId();
+            if (AWAITING_LAZY_ASSIGNMENT.equals(task.getAssignment())) {
+                extraSingleNodeModelMemoryInBytes = Math.max(
+                    extraSingleNodeModelMemoryInBytes,
+                    mlMemoryTracker.getAnomalyDetectorJobMemoryRequirement(jobId)
+                );
+                extraSingleNodeProcessors = Math.max(extraSingleNodeProcessors, 1);
+                ++extraProcessors;
+            } else {
+                modelMemoryBytesSum += mlMemoryTracker.getAnomalyDetectorJobMemoryRequirement(jobId);
+                nodesWithRunningJobs.add(task.getExecutorNode());
+                minNodes = 1;
```

Comment:
Should we set `minNodes` to 1 even if the only anomaly detection jobs are unassigned?